src/basic/cgroup-util.c

```diff
@@ -343,7 +352,7 @@ static int cg_kill_items(
                 for (;;) {
                         _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
 
-                        r = cg_read_pidref(f, &pidref);
+                        r = cg_read_pidref(f, &pidref, /* flags = */ 0);
```

Comment:
So this is actually broken: by setting flags==0, we allow cg_read_pidref to return pid==0, which is turned in reference to $self by cg_read_pid().

The approach with flags is painfully verbose but also very hard to understand, because it creates an action at a distance, where if don't set the flag, we may suddenly get a value that the code wasn't expecting before. I'll submit an alternate PR that takes a different approach.

Reply:
Hmm, OK. I pushed https://github.com/systemd/systemd/pull/32722 which I think is a simpler solution that doesn't create too much code churn. Please push your version, and then we can see what works better.